README.md
<div>
    <a href="https://travis-ci.com/kiwiroy/mojo-transaction-http-role-mechanize">
      <img alt="Travis Build Status" src="https://travis-ci.com/kiwiroy/mojo-transaction-http-role-mechanize.svg?branch=master" />
    </a>
    <a href="https://kritika.io/users/kiwiroy/repos/7509235145731088/heads/master/">
      <img alt="Kritika Analysis Status"
           src="https://kritika.io/users/kiwiroy/repos/7509235145731088/heads/master/status.svg?type=score%2Bcoverage%2Bdeps" />
    </a>
    <a href="https://coveralls.io/github/kiwiroy/mojo-transaction-http-role-mechanize?branch=master">
      <img alt="Coverage Status"
           src="https://coveralls.io/repos/github/kiwiroy/mojo-transaction-http-role-mechanize/badge.svg?branch=master" />
    </a>
</div>

# NAME

Mojo::Transaction::HTTP::Role::Mechanize - Mechanize Mojo a little

# SYNOPSIS

    # description
    my $tx = $ua->get('/')->with_roles('+Mechanize');
    my $submit_tx = $tx->submit('#submit-id', username => 'fry');
    $ua->start($submit_tx);

# DESCRIPTION

[Role::Tiny](https://metacpan.org/pod/Role::Tiny) based role to compose a form submission _"trait"_ into
[Mojo::Transaction::HTTP](https://metacpan.org/pod/Mojo::Transaction::HTTP).

# METHODS

[Mojo::Transaction::HTTP::Role::Mechanize](https://metacpan.org/pod/Mojo::Transaction::HTTP::Role::Mechanize) implements the following methods.

## submit

    # result
    $submit_tx = $tx->submit('#id', username => 'fry');

Build a new [Mojo::Transaction::HTTP](https://metacpan.org/pod/Mojo::Transaction::HTTP) object with
["tx" in Mojo::UserAgent::Transactor](https://metacpan.org/pod/Mojo::UserAgent::Transactor#tx) and the contents of the `form` with the
`$id` and merged values.

# AUTHOR

kiwiroy - Roy Storey <kiwiroy@cpan.org>

# CONTRIBUTORS

tekki - Rolf Stöckli <tekki@cpan.org>

# LICENSE

This library is free software and may be distributed under the same terms as
perl itself.


Changes
Revision history for perl distribution Mojo-Transaction-HTTP-Role-Mechanize

0.06 Not Released
 - Use correct css selector for submits #9, #10

0.05 2020-02-29T08:57:41+1300
 - added extract_forms method #7

0.04 2019-07-26T23:26:42+1200
 - tx->submit accept list and hash #5, #6

0.03 2019-07-25T20:43:21+1200
 - Ensure inputs with different type, but same name are included #3
 - Unify the way default buttons are used

0.02 2019-07-24T11:05:55+1200
 - Declare dependencies correctly #1
 - remove signatures for now #2

0.01 2019-07-23T18:57:28+1200
 - Started project from mojolicious/mojo#1387


Makefile.PL
# Generated by git-ship. See 'git-ship --man' for help or https://github.com/jhthorsen/app-git-ship
use utf8;
use strict;
use warnings;
use ExtUtils::MakeMaker;
my %WriteMakefileArgs = (
  NAME           => 'Mojo::Transaction::HTTP::Role::Mechanize',
  AUTHOR         => 'Roy Storey <kiwiroy@users.noreply.github.com>',
  LICENSE        => 'artistic_2',
  ABSTRACT_FROM  => 'lib/Mojo/Transaction/HTTP/Role/Mechanize.pm',
  VERSION_FROM   => 'lib/Mojo/Transaction/HTTP/Role/Mechanize.pm',
  EXE_FILES      => [qw()],
  BUILD_REQUIRES => {}
,
  TEST_REQUIRES  => {
  'Test::More' => '0.80'
}
,
  PREREQ_PM      => {
  'Class::Method::Modifiers' => '2.00',
  'Mojolicious' => '8.10',
  'Role::Tiny' => '2.000001'
}
,
  META_MERGE     => {
    'dynamic_config' => 0,
    'meta-spec'      => {version => 2},
    'resources'      => {
      bugtracker => {web => 'https://github.com/kiwiroy/mojo-transaction-http-role-mechanize/issues'},
      homepage   => 'https://github.com/kiwiroy/mojo-transaction-http-role-mechanize',
      repository => {
        type => 'git',
        url  => 'https://github.com/kiwiroy/mojo-transaction-http-role-mechanize.git',
        web  => 'https://github.com/kiwiroy/mojo-transaction-http-role-mechanize',
      },
    },
    'x_contributors' => []
,
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
  dist => { ($^O eq 'darwin' ? (TAR => 'gnutar') : ())},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/DOM/Role/Form.pm
package Mojo::DOM::Role::Form;

use Mojo::Base -role;

requires qw{ancestors at attr find matches selector tag val};

sub target {
  my ($self, $submit) = (shift, shift);
  return () if ($self->tag // '') ne 'form';
  return ()
    unless defined($submit = $self->at($submit || _form_default_submit($self)));
  return () if $submit->matches('[disabled]');
  my $method = uc($submit->attr('formmethod') || $self->attr('method') || 'GET');
  my $action = $submit->attr('formaction') || $self->attr('action') || '#';
  my $enctyp = $submit->attr('formenctype') || $self->attr('enctype') ||
    'url-encoded';
  return $method, $action, $enctyp;
}

around val => sub {
  my ($orig, $self, @args) = @_;
  # "form"
  return
    $self->find('button, checkbox, input, radio, select, textarea')
      ->map(sub {
        my $is_image = !!$_->matches('input[type=image]');
        # ignore disabled nodes
        return () if _form_element_disabled($_);
        # ignore those without name, unless image type
        return () if !defined(my $name = $_->attr("name")) && !$is_image;
        # only continue if the clickable element matches (synthesize click)
        return () if _form_element_submits($_) && !$_->matches($_[1]);
        # client only buttons ignored
        return () if _form_element_client_only_button($_);
        # simply return name => value for all but image types
        return [$name => $_->val()] unless $is_image;
        # synthesize image click
        return _form_image_click($_, $name);
      }, $args[0] || _form_default_submit($self))
      ->reduce(sub {
        my ($key, $value) = @$b;
        $a->{$key} = defined $a->{$key} && defined($value)
          ? [ ref($a->{$key}) ? (@{$a->{$key}}, $value) : ($a->{$key}, $value) ]
          : $value;
        $a
      }, {})
  if (my $tag = $self->tag) eq 'form';

  # "option"
  return $self->{value} // $self->text if $tag eq 'option';

  # "input" ("type=checkbox" and "type=radio")
  my $type = $self->{type} // '';
  return $self->{value} // 'on'
    if $tag eq 'input' && ($type eq 'radio' || $type eq 'checkbox');

  # "textarea", "input" or "button". Give input[type=submit] default value
  return (
    $tag eq 'textarea'
    ? $self->text
    : ($self->matches('input[type=submit]')
      ? $self->{value} || 'Submit'
      : $self->{value})) if $tag ne 'select';

  # "select"
  my $v = $self->find('option:checked:not([disabled])')
    ->grep(sub { !$_->ancestors('optgroup[disabled]')->size })->map('val');
  return exists $self->{multiple} ? $v->size ? $v->to_array : undef : $v->last;
};

#
# internal
#

sub _form_default_submit {
  # filter for those submittable nodes
  return shift->find('*')->grep(sub { !!$_->_form_element_submits; })
    # only the first continues, save some cycles
    ->tap(sub { splice @$_, 1; })
    ->map(sub {
      # $_->selector;
      # get the selector and relativise to form
      (my $s = $_->selector) =~ s/^.*form[^>]*>\s//;
      return $s;
    })->first || '';
}

sub _form_element_client_only_button {
  my $s = 'input[type=button], button[type=button], button[type=reset]';
  return !!$_[0]->matches($s);
}

sub _form_element_disabled {
  return 1 if $_[0]->matches('[disabled]');
  return 1 if $_[0]->ancestors('fieldset[disabled]')->size &&
    !$_[0]->ancestors('fieldset legend:first-child')->size;
  return 0;
}

sub _form_element_submits {
  my $s = join ', ', 'button:not([type=button], [type=reset])',
    'button', # submit is the default
    'input[type=submit]', 'input[type=image]';
  return 1 if $_[0]->matches($s) && !_form_element_disabled($_[0]);
  return 0;
}

sub _form_image_click {
  my ($self, $name) = (shift, shift);
  my ($x, $y) = map { int(rand($self->attr($_) || 1)) + 1 } qw{width height};
  # x and y if no name
  return ([x => $x], [y => $y]) unless $name;
  # named x and y, with name
  return (["$name.x" => $x], ["$name.y" => $y]);
}

1;

=encoding utf8

=head1 NAME

Mojo::DOM::Role::Form - Form data extraction

=head1 SYNOPSIS

  # description
  my $obj = Mojo::DOM::Role::Form->new();
  $obj->target('#submit-id');

=head1 DESCRIPTION

L<Role::Tiny> based role to compose additional form data extraction methods into
L<Mojo::DOM>.

=head1 METHODS

L<Mojo::DOM::Role::Form> implements the following methods.

=head2 target

  # result
  $obj->target 

Explain what the L</"target"> does.

=head1 AUTHOR

=cut


lib/Mojo/Transaction/HTTP/Role/Mechanize.pm
package Mojo::Transaction::HTTP::Role::Mechanize;

use Mojo::Base -role;
use Mojo::UserAgent::Transactor;

our $VERSION = '0.05';

requires qw{error res};

sub extract_forms {
  my $self = shift;

  my $forms
    = $self->res->dom->find('form')->each(sub { $_->with_roles('+Form') });

  return $forms;
}

sub submit {
  my ($self, $selector, $overlay) = (shift);
  $overlay  = pop   if @_ && ref($_[-1]) eq 'HASH';
  $selector = shift if @_ % 2;
  $overlay ||= {@_};

  # cannot continue from error state
  return if $self->error;

  # extract form
  my $form = $self->extract_forms->grep(sub { $_->at($selector // '') })->first
    or return;

  return unless (my ($method, $target, $type) = $form->target($selector));
  $target = $self->req->url->new($target);
  $target = $target->to_abs($self->req->url) unless $target->is_abs;

  # values from form
  my $state = $form->val($selector);

  # merge in new values of form elements
  my @keys = grep { exists $overlay->{$_} } keys %$state;
  @$state{@keys} = @$overlay{@keys};

  # build a new transaction ...
  return Mojo::UserAgent::Transactor->new->tx(
    $method => $target,
    {}, form => $state
  )->previous($self);
}

1;

=encoding utf8

=begin html

<img alt="Build Status - Linux"
     src="https://github.com/kiwiroy/mojo-transaction-http-role-mechanize/workflows/linux/badge.svg" />
<img alt="Build Status - Macos"
    src="https://github.com/kiwiroy/mojo-transaction-http-role-mechanize/workflows/macos/badge.svg" />
<img alt="Build Status - Windows"
     src="https://github.com/kiwiroy/mojo-transaction-http-role-mechanize/workflows/windows/badge.svg" />
<a href="https://kritika.io/users/kiwiroy/repos/7509235145731088/heads/master/">
  <img alt="Kritika Analysis Status"
       src="https://kritika.io/users/kiwiroy/repos/7509235145731088/heads/master/status.svg?type=score%2Bcoverage%2Bdeps" />
</a>
<a href="https://coveralls.io/github/kiwiroy/mojo-transaction-http-role-mechanize?branch=master">
  <img alt="Coverage Status"
       src="https://coveralls.io/repos/github/kiwiroy/mojo-transaction-http-role-mechanize/badge.svg?branch=master" />
</a>
<a href="https://badge.fury.io/pl/Mojo-Transaction-HTTP-Role-Mechanize">
  <img alt="CPAN version"
       src="https://badge.fury.io/pl/Mojo-Transaction-HTTP-Role-Mechanize.svg" />
</a>

=end html

=head1 NAME

Mojo::Transaction::HTTP::Role::Mechanize - Mechanize Mojo a little

=head1 SYNOPSIS

  use Mojo::UserAgent;
  use Mojo::Transaction::HTTP::Role::Mechanize;

  my $ua = Mojo::UserAgent->new;
  my $tx = $ua->get('/')->with_roles('+Mechanize');

  # call submit immediately
  my $submit_tx = $tx->submit('#submit-id', username => 'fry');
  $ua->start($submit_tx);

  # first extract form values
  my $values = $tx->extract_forms->first->val;
  $submit_tx = $tx->submit('#submit-id', counter => $values->{counter} + 3);
  $ua->start($submit_tx);

=head1 DESCRIPTION

L<Role::Tiny> based role to compose a form submission I<"trait"> into
L<Mojo::Transaction::HTTP>.

=head1 METHODS

L<Mojo::Transaction::HTTP::Role::Mechanize> implements the following method.

=head2 extract_forms

  $collection = $tx->extract_forms;

Returns a L<Mojo::Collection> of L<Mojo::DOM> elements with activated L<Mojo::DOM::Role::Form>
that contains all the forms of the page.

=head2 submit

  # result using selector
  $submit_tx = $tx->submit('#id', username => 'fry');

  # result without selector using default submission
  $submit_tx = $tx->submit(username => 'fry');

  # passing hash, rather than list, of values
  $submit_tx = $tx->submit({username => 'fry'});

  # passing hash, rather than list, of values and a selector
  $submit_tx = $tx->submit('#id', {username => 'fry'});

Build a new L<Mojo::Transaction::HTTP> object with
L<Mojo::UserAgent::Transactor/"tx"> and the contents of the C<form> with the
C<$id> and merged values.  If no selector is given, the first non-disabled
button or appropriate input element (of type button, submit, or image)
will be used for the submission.

=head1 AUTHOR

kiwiroy - Roy Storey C<kiwiroy@cpan.org>

tekki - Rolf Stöckli C<tekki@cpan.org>

lindleyw - William Lindley C<wlindley+remove+this@wlindley.com>

=head1 LICENSE

This library is free software and may be distributed under the same terms as
perl itself.

=head1 SEE ALSO

L<Mojo::DOM::Role::Form>, L<Mojolicious>.

=cut
